pkg/controller/job/job_controller.go

```diff
@@ -727,7 +741,22 @@ func (jm *Controller) manageJob(activePods []*v1.Pod, succeeded int32, job *batc
 			}(i)
 		}
 		wait.Wait()
+	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && job.Spec.Suspend {
```

Comment:
A new test case in `TestControllerSyncJob` should uncover all the edge cases, at least I'd like to see:
1. active job with some pods in various states being suspended (w/o required pod change, for example parallelism == activePods)
2. active job with some pods in various states being suspended (with pod change required, parallelism != activePods)
3. new job being suspended
4. deleted job

Reply:
>b/c it relies that a job actually needs a change

That's a great point, thank you for catching that! Isn't this achieved by checking for suspension state change in addition to jobNeedsSync?

>There's a question what happens if start time isn't set (a new suspended job) should we set it or keep it nil until it actually starts?

We decided in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2232-suspend-jobs#design-details) that:

>If a Job is created with `suspend: true`, the `StartTime` field of the Job status is set only when it is resumed for the first time.

I agree with @ahg-g about ActiveDeadlineSeconds. The way I've updated the code, the timer will now only be set after StartTime is set, which will happen only after the job is unsuspended for the first time i.e. jobs created in the suspended state can stay that way for longer than ActiveDeadlineSeconds.

>Parallelism is a mutable field, and users can set this field back to 0; what I failed to understand when reading the code is how a job update that sets parallelism=0 result in a job sync (i.e., a call to manageJob) because I assume we can do the same for when suspend is changed, and so we keep the code pretty much as is.

@ahg-g Setting parallelism=0 will normally trigger the `active > parallelism` condition below, but we get a similar behaviour by checking for suspend before that condition. The suspend branch (this branch) does the same thing as the `active > parallelism` branch when `parallelism` is 0 through the `deletePods` function. Functionally they're similar, but they're simply executed in different branches so that the log message etc. could be different. Unless I misunderstood your question?

Reply:
I had a chat with all the parties in the slack, the confusion was coming from my initial different take on how we should approach job suspension, which I thought will be like a hammer, but we eventually decided it will not, but it'll be treated equally with any other change to a job spec (like @ahg-g mentioned above parallelism, for example). 

So the only remaining thing is changes wrt ADS and unit tests.

Reply:
Done.